braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,1071 @@
+// Include necessary headers
+#define _GNU_SOURCE
+#include <strings.h>
+#include <cupsfilters/log.h>
+#include <cupsfilters/filter.h>
+#include <ppd/ppd-filter.h>
+#include <limits.h>
+#include <pappl/pappl.h>
+#include <liblouisutdml/liblouisutdml.h>
+#include <cups/backend.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <string.h>
+#include <cups/ipp.h>
+#include <magic.h>
+
+#include "brf-printer.h"
+
+// Include necessary headers...
+
+#define brf_TESTPAGE_MIMETYPE "application/vnd.cups-brf"
+
+extern bool brf_gen(pappl_system_t *system, const char *driver_name, const char *device_uri, const char *device_id, pappl_pr_driver_data_t *data, ipp_t **attrs, void *cbdata);
+extern char *strdup(const char *);
+
+static bool BRFTestFilterCB(pappl_job_t *job, pappl_device_t *device, void *cbdata);
+
+static int brf_print_filter_function(int inputfd, int outputfd, int inputseekable, cf_filter_data_t *data, void *parameters);
+
+static const char *autoadd_cb(const char *device_info, const char *device_uri, const char *device_id, void *cbdata);
+
+static bool driver_cb(pappl_system_t *system, const char *driver_name, const char *device_uri, const char *device_id, pappl_pr_driver_data_t *data, ipp_t **attrs, void *cbdata);
+
+static int match_id(int num_did, cups_option_t *did, const char *match_id);
+
+static const char *mime_cb(const unsigned char *header, size_t headersize, void *data);
+
+static bool printer_cb(const char *device_info, const char *device_uri, const char *device_id, pappl_system_t *system);
+
+static pappl_system_t *system_cb(int num_options, cups_option_t *options, void *data);
+static int create_brf_printer(pappl_system_t *system);
+
+// Local globals...
+
+static pappl_pr_driver_t brf_drivers[] =
+    {
+        // Driver list
+        {"gen_brf", "Generic Braille embosser",
+         NULL, NULL},
+
+};
+
+// State file
+
+static char brf_statefile[1024];
+
+// 'main()' - Main entry for brf.
+
+int                // O - Exit status
+main(int argc,     // I - Number of command-line arguments
+     char *argv[]) // I - Command-line arguments
+{
+
+  cups_array_t *spooling_conversions;
+  spooling_conversions = cupsArrayNew(NULL, NULL);
+
+  for (int i = 0; converts[i].srctype != NULL; i++)
+  {
+    cupsArrayAdd(spooling_conversions, &converts[i]);
+  }
+
+  brf_printer_app_config_t printer_app_config = {
+      .spooling_conversions = spooling_conversions};
+
+  brf_printer_app_global_data_t global_data;
+
+  memset(&global_data, 0, sizeof(global_data));
+
+  global_data.config = &printer_app_config;
```

Comment:
You can as well make `config` directly a `brf_printer_app_config_t`, not a `brf_printer_app_config_t *`, so you don't need to set this pointer (and not even need to call memset since the `global_data` variable will already be initialized to 0)